server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -566,23 +566,21 @@ private static void parseDynamicValue(final ParseContext context, ObjectMapper p
         }
         final Mapper.BuilderContext builderContext = new Mapper.BuilderContext(context.indexSettings().getSettings(), context.path());
         final Mapper.Builder<?> builder = createBuilderFromDynamicValue(context, token, currentFieldName);
-        if (parentMapper.equals(context.root())) {
-            int position = getPositionForDynamicField(context, parentMapper);
-            if (builder instanceof FieldMapper.Builder<?> fieldMapperBuilder) {
-                fieldMapperBuilder.position(position);
-            }
+        int position = getPositionForDynamicField(context);
+        if (builder instanceof FieldMapper.Builder<?> fieldMapperBuilder) {
+            fieldMapperBuilder.position(position);
```

Comment:
Can you explain the problem with the previous approach?

Reply:
```
CREATE TABLE IF NOT EXISTS "doc"."testing" (  
           "p1" text,  
           "nested" OBJECT(DYNAMIC) AS (  
              "sub1" OBJECT(DYNAMIC) AS (  
                 "sub2" BIGINT  
                )  
           ),  
           "obj" OBJECT(DYNAMIC)  
        );                        
        
INSERT INTO "doc"."testing" ("p1", "obj") VALUES ('a',  '{"new2": 1}');
```

In this case, the parent mapper is `obj` when inserting `new2`. So, in order to assign the position I think the `if` should be removed.